Reduce WaitForNoFollowersStep requests indices shard stats

luyuncheng opened the pull request:
When data is time series indices, it may concurrent create many index and delete index using `ILM`

i see `DeleteAction`, `ShrinkAction` need to check ccr status in `WaitForNoFollowersStep` as following code shows:

https://github.com/elastic/elasticsearch/blob/39625567e0fb5acf8d8d5f7a30bd17e4b7e2fab3/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DeleteAction.java#L81-L96

`WaitForNoFollowersStep` do collect indices stats to evaluate the condition
https://github.com/elastic/elasticsearch/blob/39625567e0fb5acf8d8d5f7a30bd17e4b7e2fab3/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForNoFollowersStep.java#L56-L63

BUT indices stats api need broadcast to many nodes and collect `shardStats` which is time consuming, even we know the CCR is disabled

![image](https://user-images.githubusercontent.com/12760367/225207020-f7f5fcd4-9f55-48c8-abfb-fce3d1f0dcfd.png)

i think, we can use feature api get feature info, and reduce indices stats request when the CCR is disabled

luyuncheng committed:
1. Reduce WaitForNoFollowersStep requests indices shard stats

luyuncheng commented:
@andreidan i know there are many optimizations in ilm batch task AND master batch queue like 78547, 92021, 94325 etc. BUT when many deleteAction triggered in ILM, all phrase would wait in `WaitForNoFollowersStep` because it should do collect shard stats from master which is time consuming, and other phase can batch execute.

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForNoFollowersStep.java:
```diff
@@ -49,10 +53,31 @@ public boolean isRetryable() {
 
     @Override
     public void evaluateCondition(Metadata metadata, Index index, Listener listener, TimeValue masterTimeout) {
+        XPackInfoRequest xPackInfoRequest = new XPackInfoRequest();
```
We have a bit of precedent here as we already skip the `UnfollowAction` completely if no CCR information is present in the index metadata.
See https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/UnfollowAction.java#L62 

How about we do the same here? (we have the `Metadata` and the `Index` so we can use a similar check here and if the `ccr` metadata is missing  for this index we notify `listener.onResponse(true)`) 

This avoids an extra API transport call.

luyuncheng replied:
@andreidan As i read the code,  
`UnfollowAction` just check in the following cluster, and  there is a `ccr` KEY  in the metadata 
BUT
`WaitforNoFollowersStep` check in the leader cluster to ensure policy run safely that all followers stoped, and introduced in #38140.  There is no `ccr` KEY in metadata statement at leader cluster. so we can not do the same as `UnfollowAction`. 

The leader cluster get the CCR status only from the shards stats in `CCR_LEASE_KEY` AND feature API which CCR is disabled

luyuncheng replied:
> > There is no ccr KEY in metadata statement at leader cluster
> 
> Ah of course ! Good shout @luyuncheng

@andreidan  i think `XPackInfoRequest` is a more lightweight calling VS `shards stats` requests. It can reduce master AND data nodes calculating, memory pressure without collect shards stats

andreidan replied:
++ I think it makes sense, thanks for looking into this

andreidan reviewed (commented):
Thanks for looking into this @luyuncheng

I left an alternative solution suggestion. Hope it makes sense.

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/WaitForNoFollowersStep.java:
```diff
@@ -49,10 +53,31 @@ public boolean isRetryable() {
+                XPackInfoResponse.FeatureSetsInfo featureSetsInfo = xPackInfoResponse.getFeatureSetsInfo();
+                if (featureSetsInfo != null) {
+                    XPackInfoResponse.FeatureSetsInfo.FeatureSet featureSet = featureSetsInfo.getFeatureSets().get(CCR_LEASE_KEY);
+                    if (featureSet != null && (featureSet.available() == false || featureSet.enabled() == false)){
```
Let's please just check if the featureSet is enabled here.
e.g. 
```
if (featureSet != null && featureSet.enabled() == false)
```

We'd like to maintain the same behaviour for this step even if CCR is unavailable (e.g. license is expired)

luyuncheng replied:
> We'd like to maintain the same behaviour for this step even if CCR is unavailable (e.g. license is expired)

LGTM

andreidan commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/WaitForNoFollowersStepTests.java:
```diff
@@ -64,7 +68,7 @@ protected WaitForNoFollowersStep copyInstance(WaitForNoFollowersStep instance) {
     }
 
-    public void testConditionMet() {
+    public void testConditionMetWhenEnable() {
```
Shall we name this `testConditionMetWhenCCREnabled` ?

luyuncheng replied:
++

andreidan commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/WaitForNoFollowersStepTests.java:
```diff
@@ -76,6 +80,41 @@ public void testConditionMet() {
+        assertNull(stepInfoHolder.get());
+    }
+
+    public void testConditionMetWhenDisabled() {
```
Similarly, shall we name this test to include `CCRDisabled` ?

andreidan reviewed (changes requested):
Thanks for working on this @luyuncheng 

I've left a few rather minor suggestions (and let's see if CI is happy)

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into WaitForNoFollowersStep

luyuncheng committed:
1. format && rename test name

luyuncheng committed:
1. maintain the same behaviour for this step even if CCR is unavailable (e.g. license is expired)

luyuncheng committed:
1. Use specific ccr FeatureAction instead of XPackInfoAction
2. Add test dependency ccr for ilm
3. Add InternalClusterTests for CCR disabled

luyuncheng commented:
The  previous ci test https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request+part-2/33165/ fail in `intenalClusterTest` because `ActionType` not found. i think this because `XpackInfoFeatureAction` scan all xpack module like following code shows:

https://github.com/elastic/elasticsearch/blob/a2c9c0b82f3c12aa5988d8f643aec53f024f539a/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/action/TransportXPackInfoAction.java#L36-L51

and internalClusterTest in ILM not include other module into plugins when override `nodePlugins`:

https://github.com/elastic/elasticsearch/blob/a2c9c0b82f3c12aa5988d8f643aec53f024f539a/x-pack/plugin/ilm/src/internalClusterTest/java/org/elasticsearch/xpack/ilm/ClusterStateWaitThresholdBreachTests.java#L69-L72

in commits https://github.com/elastic/elasticsearch/pull/94510/commits/d3c5f990f837d1d50c85f526ce2ef108c29c31bb  i added dependency ccr into build.gradle. and i also added internalClusterTest to check this

andreidan commented on x-pack/plugin/ilm/src/internalClusterTest/java/org/elasticsearch/xpack/ilm/ILMMultiNodeWithCCREnabledIT.java:
```diff
@@ -41,12 +43,12 @@
 
 @ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0, numClientNodes = 0)
-public class ILMMultiNodeIT extends ESIntegTestCase {
+public class ILMMultiNodeWithCCREnabledIT extends ESIntegTestCase {
```
I think we don't have to rename this as it's not meant to test anything CCR related

luyuncheng replied:
++

andreidan commented on x-pack/plugin/ilm/src/internalClusterTest/java/org/elasticsearch/xpack/ilm/ILMMultiNodeWithCCRDisabledIT.java:
```diff
@@ -0,0 +1,136 @@
+        startWarmOnlyNode();
+        ensureGreen();
+
+        RolloverAction rolloverAction = new RolloverAction(null, null, null, 1L, null, null, null, null, null, null);
```
I think we can skip the hot phase completely to save a few cycles in this test

luyuncheng replied:
> I think we can skip the hot phase completely to save a few cycles in this test

I need `ShrinkAction` to trigger `WaitForNoFollowersStep` and shrink action need a read-only index. so i combine the `RolloverAction` which is for enable read-only index AND `ShrinkAction` into one hot phase. and  **skip** the warm phase

andreidan reviewed (commented):
Thanks for iterating on this @luyuncheng 

Left a few minor comments. 

Can you please also add a changelog entry ? (similar to this https://github.com/elastic/elasticsearch/blob/main/docs/changelog/93524.yaml )

luyuncheng committed:
Add changelog AND modified tests

luyuncheng committed:
Add changelog AND modified tests AND spotless check

luyuncheng committed:
Merge pull request #4 from luyuncheng/master

update master branch

luyuncheng commented:
> Can you please also add a changelog entry ? 

++

andreidan commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into WaitForNoFollowersStep

andreidan reviewed (approved):
LGTM, thanks for the contribution @luyuncheng
